---
title: "Midterm: Group comparison analysis"
author: "Aizhan Uteubayeva (NetID:au198)"
date: "2023-10-12"
output: html_document
---
### 1a - Read in clinical data

```{r}
## 22 patients
clinData <- read.csv(file = "input/Blalock_clin_final.csv",
                       sep=",", 
                       header = T)
knitr::kable(head(clinData[,1:4],n = 5))
```
### 1b: Read in processed normalized gene expression data in log2 scale, includes gene annotation

```{r}

# Read in gene expressin file - 41927 feature (rows), 91 patients (columns)
geneExp <- read.table(file = "GSE62232_Blalock_geneexp_final.tsv",
                      sep="\t",
                      row.names = 1,
                      header = T, 
                      stringsAsFactors = F)

knitr::kable(geneExp[1:5, 1:4], caption = "Processed gene expression data in log2 scale with gene annotation")
```
### The Goal is to compare which two groups ?
The goal is to compare the contral vs. incipient patients
```{r}
knitr::kable(table(clinData$DISEASE_STATUS))
```
## Step 2 - Clean/Filter data

### Filter clinical data - not needed

Clinical data has one row per patient. So cleaned and ready to go

```{r}
dim(clinData)
```
### Filter gene exp data

Check rows and columns:

-   Rows are features (genes in this case)
-   Columns are patients

```{r}
knitr::kable(geneExp[1:5, 1:4]) 
```
### What are the sample identifiers ?

The GSM ids are the identifiers in the gene exp data that match the identifiers for the Clinical data file:

-   Sample identifier for clinical data: Column called "BIOSPECIMEN_ID"
-   Sample identifier for gene expression data: Column names are the GSMIds

```{r}
matchingSamples <- which(colnames(geneExp) %in% clinData$BIOSPECIMEN_ID) # 30 IDs matching
subsetGeneExp <- geneExp[, matchingSamples] ## 43135 rows features, 30 patient columns
# ALL IDS HAVE MATCHED
```
## Step 3 - Identify the groups to be compared

### Identify the groups to be compared (Baseline and Comparison Grps)

```{r}
# Labels (row numbers) that can identify the baseline group patients
baselineGrpLabels <-  which(clinData$DISEASE_STATUS == "Control") #8 samples
head(baselineGrpLabels)
# Use the labels (row numbers) to subset baseline patients in clinical data file 
clinBase <- clinData[baselineGrpLabels, ]

# Labels (row numbers) that can identify the comp group patients
compGrpLabels <- which(clinData$DISEASE_STATUS == "Incipient") #7 samples
head(compGrpLabels)
# Use the labels (row numbers) to subset comp patients in clinical data file 
clinComp <- clinData[compGrpLabels, ]

#### Use the clinBase and clinComp objects to subset gene expression data
geneExpAlzBase <- subsetGeneExp[, clinBase$BIOSPECIMEN_ID] # 43135 feature (rows), 8 samples columns
geneExpAlzComp <- subsetGeneExp[, clinComp$BIOSPECIMEN_ID] # 43135 feature (rows), 7 samples columns
```
## Step 4: Sanity check

-   See if filtering of clinical data in R matches filtering of clinical data in excel
-   See if sample ids in clinical data match sample ids in gene exp data (if they don't match it means your step 1 and/or 2 is wrong)
-   Verify you see correct number of samples in baseline and comp groups
-   Export the column names from gene expression data to see if it contains only probe/gene names and no other garbage

```{r}
#See if sample ids in clinical data match sample ids in gene exp data
clinBase$BIOSPECIMEN_ID == colnames(geneExpAlzBase)
clinComp$BIOSPECIMEN_ID == colnames(geneExpAlzComp)

### Checking if the samples IDs baseline and comp groups are the same ---- you can put these in an excel to see if IDs match
write.table(clinBase$BIOSPECIMEN_ID , file = "output/ClinBaseIDs.tsv", sep="\t", quote = F  )
write.table(clinComp$BIOSPECIMEN_ID , file = "output/ClinCompIDs.tsv", sep="\t", quote = F  )
write.table(colnames(geneExpAlzBase) , 
            file = "output/GeneExpBaseIDs.tsv", 
            sep="\t", quote = F  )
write.table(colnames(geneExpAlzComp) , 
            file = "output/GeneExpCompIDs.tsv", 
            sep="\t", quote = F  )

#Export the features from gene expression data  
#Open this file and check that it contains only probe/gene names and no other garbage
write.table(rownames(geneExp),file = "output/genExpFeatureIDsCheck.tsv", sep="\t", quote = F  )
```
## Step 5: Prep data for T-test

-   Molecular data must have features (genes in this case) as rows, and samples as columns.
-   Transpose data (if needed) to obtain this
-   Objects must be data frame
-   Numeric data only

```{r}
### Checking to make sure data is a numeric data frame
knitr::kable(head(geneExpAlzBase[1:5,1:4]))
knitr::kable(head(geneExpAlzComp[1:5,1:4]))
```
#### Function for T-test

```{r ttest}
source("input/fnTTest.R")

#### Call T-test function
results1 = fnTTest(baseGroup = geneExpAlzBase, 
                   compGroup = geneExpAlzComp, 
                   testName = "Alzheimer's", 
                   baseGroupName = "Control",
                   compGroupName =  "Incipient",
                   folderName = "output")
```
*The output of the T-Test analysis is a CSV File , and will be located in the folderName that you provided in the previous step. Locate this file.*

# Next Step - Sub-set top differentially expressed genes

Clear your workspace Read in the T-Test results file

```{r}
#Read in the T-Test results file
ttestResults <- read.csv(file = "output/Alzheimer's_Incipient_(Comp).vs._Control_(Base).TTest.csv")

#check to make sure p-value column is imported as numeric 
#sort by p-value (just in case the results are not sorted by p-value)
ttestResultsSorted <- dplyr::arrange(ttestResults, Pvalue)

#find rows with p-value < 0.01
whichSig <- which(ttestResultsSorted$Pvalue <= 0.01) 

#Short list sig results
ttestResultsSig <- ttestResultsSorted[whichSig, ] #294 rows 

### Export short listed results
write.table(x = ttestResultsSig, 
            file = "output/Midterm_Alzheimers_Ttest_Shortlisted.tsv", 
            quote = F, sep = "\t")

##### First column is a list of features in thsi format : ProbeID|GeneName. 
#### Use string split strsplit() function to extract gene names
funcSplit <- function(featureX) {
  f1 <- unlist(strsplit(x = featureX, split = "|", fixed = TRUE))
  f2 <- f1[2]
  return(f2)
}

# Use apply() function to run the split on every row, its faster version of a loop
geneNames1 <- apply(X = as.matrix(ttestResultsSig$Feature), 
                    MARGIN = 1, FUN = funcSplit)

head(geneNames1)

unique_genes <- unique(geneNames1)

#print length of short listed gene names
length(unique_genes)

### Export list of gene names
write.table(x = unique_genes, 
            file = "output/Unique_Midterm_Alzheimers_SigDiffExpressedGenes.tsv", 
            quote = F, sep = "\t")
```

